crates/ty_python_semantic/src/types/signatures.rs

```diff
@@ -400,10 +403,75 @@ impl<'db> Signature<'db> {
         }
     }
 
+    /// Creates a new version of this signature with its generic context specialized to match
+    /// another signature's generic context. This is used for assignability/subtyping checks
+    /// between two generic function signatures where the type variable identity shouldn't matter.
+    fn with_specialized_generic_context<'a>(
+        &'a self,
+        db: &'db dyn Db,
+        other: &'a Signature<'db>,
+    ) -> Cow<'a, Signature<'db>> {
+        if let (Some(self_gc), Some(other_gc)) = (self.generic_context, other.generic_context) {
+            // If the generic contexts are the same, the TypeVars are already aligned.
+            // No specialization is needed.
+            if self_gc == other_gc {
+                return Cow::Borrowed(self);
+            }
+            let mut builder = SpecializationBuilder::new(db);
+
+            // If we get an error, return the original signature without specialization. All
+            // subsequent comparisons will fail.
+            if self
+                .infer_generic_specialization(db, other, &mut builder)
+                .is_err()
+            {
+                return Cow::Borrowed(self);
+            }
+
+            if let (Some(self_return), Some(other_return)) = (self.return_ty, other.return_ty) {
+                // If we get an error, return the original signature without specialization. All
+                // subsequent comparisons will fail.
+                if builder.infer(self_return, other_return).is_err() {
+                    return Cow::Borrowed(self);
+                }
+            }
+
+            let specialization = builder.build(self_gc);
+            let signature =
+                self.apply_type_mapping(db, &TypeMapping::Specialization(specialization));
+            return Cow::Owned(signature);
+        }
+        Cow::Borrowed(self)
+    }
+
+    /// This is a carbon copy of [`Self::has_relation_to`], but instead of
+    /// checking types, it performs inference.
+    fn infer_generic_specialization(
```

Comment:
Should we inline this method as I think it's used in a single place and the body only has one method call?